BuildEventsListenerRegistry corrupted with Isolated Projects and parallel configuration

### Current Behavior

I enabled parallel CC load and store (`org.gradle.configuration-cache.parallel=true`) and some tests for in-house Gradle plugins randomly started failing with the following error:
```
java.lang.NullPointerException: at index 1
  at com.google.common.collect.ObjectArrays.checkElementNotNull(ObjectArrays.java:232)
  at com.google.common.collect.ObjectArrays.checkElementsNotNull(ObjectArrays.java:222)
  at com.google.common.collect.ObjectArrays.checkElementsNotNull(ObjectArrays.java:216)
  at com.google.common.collect.ImmutableList.construct(ImmutableList.java:353)
  at com.google.common.collect.ImmutableList.copyOf(ImmutableList.java:265)
  at org.gradle.internal.build.event.DefaultBuildEventsListenerRegistry.getSubscriptions(DefaultBuildEventsListenerRegistry.java:87)
  at org.gradle.internal.cc.impl.ConfigurationCacheState.buildEventListenersOf(ConfigurationCacheState.kt:884)
  at org.gradle.internal.cc.impl.ConfigurationCacheState.writeRootBuild(ConfigurationCacheState.kt:272)
  at org.gradle.internal.cc.impl.ConfigurationCacheState.writeRootBuildState(ConfigurationCacheState.kt:171)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO$writeRootBuildStateTo$1.invokeSuspend(DefaultConfigurationCacheIO.kt:190)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO$writeRootBuildStateTo$1.invoke(DefaultConfigurationCacheIO.kt)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO$writeRootBuildStateTo$1.invoke(DefaultConfigurationCacheIO.kt)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO$writeConfigurationCacheStateWithSpecialEncoders$1.invokeSuspend(DefaultConfigurationCacheIO.kt:342)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO$writeConfigurationCacheStateWithSpecialEncoders$1.invoke(DefaultConfigurationCacheIO.kt)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO$writeConfigurationCacheStateWithSpecialEncoders$1.invoke(DefaultConfigurationCacheIO.kt)
  at org.gradle.internal.serialize.graph.CodecKt$writeWith$1$1.invokeSuspend(Codec.kt:80)
  at org.gradle.internal.serialize.graph.CodecKt$writeWith$1$1.invoke(Codec.kt)
  at org.gradle.internal.serialize.graph.CodecKt$writeWith$1$1.invoke(Codec.kt)
  at org.gradle.internal.serialize.graph.RunningKt$runWriteOperation$1.invokeSuspend(Running.kt:43)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
  at kotlin.coroutines.ContinuationKt.startCoroutine(Continuation.kt:115)
  at org.gradle.internal.serialize.graph.RunningKt.runToCompletion(Running.kt:54)
  at org.gradle.internal.serialize.graph.RunningKt.runWriteOperation(Running.kt:42)
  at org.gradle.internal.serialize.graph.CodecKt.writeWith(Codec.kt:79)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.withWriteContextFor(DefaultConfigurationCacheIO.kt:497)
  at org.gradle.internal.cc.impl.ConfigurationCacheBuildTreeIO$DefaultImpls.withWriteContextFor(ConfigurationCacheBuildTreeIO.kt:112)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.withWriteContextFor(DefaultConfigurationCacheIO.kt:90)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.writeConfigurationCacheStateWithSpecialEncoders(DefaultConfigurationCacheIO.kt:341)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.access$writeConfigurationCacheStateWithSpecialEncoders(DefaultConfigurationCacheIO.kt:90)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO$writeConfigurationCacheState$1$1.invoke(DefaultConfigurationCacheIO.kt:243)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO$writeConfigurationCacheState$1$1.invoke(DefaultConfigurationCacheIO.kt:242)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.withSharedObjectEncoderFor(DefaultConfigurationCacheIO.kt:302)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.access$withSharedObjectEncoderFor(DefaultConfigurationCacheIO.kt:90)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO$writeConfigurationCacheState$1.invoke(DefaultConfigurationCacheIO.kt:242)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO$writeConfigurationCacheState$1.invoke(DefaultConfigurationCacheIO.kt:241)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.withStringEncoderFor(DefaultConfigurationCacheIO.kt:290)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.writeConfigurationCacheState(DefaultConfigurationCacheIO.kt:241)
  at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.writeRootBuildStateTo(DefaultConfigurationCacheIO.kt:188)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache.writeConfigurationCacheState(DefaultConfigurationCache.kt:504)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache.access$writeConfigurationCacheState(DefaultConfigurationCache.kt:72)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache$saveWorkGraph$1.invoke(DefaultConfigurationCache.kt:421)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache$saveWorkGraph$1.invoke(DefaultConfigurationCache.kt:420)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache$saveToCache$1$stateStoreResult$1.invoke(DefaultConfigurationCache.kt:438)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache$saveToCache$1$stateStoreResult$1.invoke(DefaultConfigurationCache.kt:435)
  at org.gradle.internal.cc.impl.ConfigurationCacheRepository$StoreImpl$useForStore$1.invoke(ConfigurationCacheRepository.kt:233)
  at org.gradle.internal.cc.impl.ConfigurationCacheRepository$StoreImpl$useForStore$1.invoke(ConfigurationCacheRepository.kt:224)
  at org.gradle.internal.cc.impl.ConfigurationCacheRepository$withExclusiveAccessToCache$1.get(ConfigurationCacheRepository.kt:291)
  at org.gradle.cache.internal.LockOnDemandCrossProcessCacheAccess.withFileLock(LockOnDemandCrossProcessCacheAccess.java:90)
  at org.gradle.cache.internal.DefaultCacheCoordinator.withFileLock(DefaultCacheCoordinator.java:213)
  at org.gradle.cache.internal.DefaultPersistentDirectoryStore.withFileLock(DefaultPersistentDirectoryStore.java:147)
  at org.gradle.cache.internal.DefaultCacheFactory$ReferenceTrackingCache.withFileLock(DefaultCacheFactory.java:203)
  at org.gradle.internal.cc.impl.ConfigurationCacheRepository.withExclusiveAccessToCache(ConfigurationCacheRepository.kt:289)
  at org.gradle.internal.cc.impl.ConfigurationCacheRepository.access$withExclusiveAccessToCache(ConfigurationCacheRepository.kt:52)
  at org.gradle.internal.cc.impl.ConfigurationCacheRepository$StoreImpl.useForStore(ConfigurationCacheRepository.kt:224)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache$saveToCache$1.invoke(DefaultConfigurationCache.kt:435)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache$saveToCache$1.invoke(DefaultConfigurationCache.kt:434)
  at org.gradle.configurationcache.ConfigurationCacheBuildOperationsKt$withStoreOperation$1.run(ConfigurationCacheBuildOperations.kt:62)
  at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
  at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
  at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
  at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
  at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:166)
  at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
  at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
  at org.gradle.configurationcache.ConfigurationCacheBuildOperationsKt.withStoreOperation(ConfigurationCacheBuildOperations.kt:55)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache.saveToCache(DefaultConfigurationCache.kt:434)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache.saveWorkGraph(DefaultConfigurationCache.kt:420)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache.access$saveWorkGraph(DefaultConfigurationCache.kt:72)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache$loadOrScheduleRequestedTasks$1.invoke(DefaultConfigurationCache.kt:173)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache$loadOrScheduleRequestedTasks$1.invoke(DefaultConfigurationCache.kt:171)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache.runWorkThatContributesToCacheEntry(DefaultConfigurationCache.kt:392)
  at org.gradle.internal.cc.impl.DefaultConfigurationCache.loadOrScheduleRequestedTasks(DefaultConfigurationCache.kt:171)
  at org.gradle.internal.cc.impl.ConfigurationCacheAwareBuildTreeWorkController$scheduleAndRunRequestedTasks$executionResult$1.apply(ConfigurationCacheAwareBuildTreeWorkController.kt:45)
  at org.gradle.internal.cc.impl.ConfigurationCacheAwareBuildTreeWorkController$scheduleAndRunRequestedTasks$executionResult$1.apply(ConfigurationCacheAwareBuildTreeWorkController.kt:44)
  at org.gradle.composite.internal.DefaultIncludedBuildTaskGraph.withNewWorkGraph(DefaultIncludedBuildTaskGraph.java:112)
  at org.gradle.internal.cc.impl.ConfigurationCacheAwareBuildTreeWorkController.scheduleAndRunRequestedTasks(ConfigurationCacheAwareBuildTreeWorkController.kt:44)
  at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$scheduleAndRunTasks$1(DefaultBuildTreeLifecycleController.java:77)
  at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$runBuild$4(DefaultBuildTreeLifecycleController.java:120)
  at org.gradle.internal.model.StateTransitionController.lambda$transition$6(StateTransitionController.java:169)
  at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:266)
  at org.gradle.internal.model.StateTransitionController.lambda$transition$7(StateTransitionController.java:169)
  at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:46)
  at org.gradle.internal.model.StateTransitionController.transition(StateTransitionController.java:169)
  at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.runBuild(DefaultBuildTreeLifecycleController.java:117)
  at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:77)
  at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:72)
  at org.gradle.tooling.internal.provider.runner.BuildModelActionRunner.run(BuildModelActionRunner.java:53)
  at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
  at org.gradle.internal.buildtree.ProblemReportingBuildActionRunner.run(ProblemReportingBuildActionRunner.java:49)
  at org.gradle.launcher.exec.BuildOutcomeReportingBuildActionRunner.run(BuildOutcomeReportingBuildActionRunner.java:66)
  at org.gradle.tooling.internal.provider.FileSystemWatchingBuildActionRunner.run(FileSystemWatchingBuildActionRunner.java:140)
  at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:41)
  at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.lambda$execute$0(RootBuildLifecycleBuildActionExecutor.java:54)
  at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:130)
  at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.execute(RootBuildLifecycleBuildActionExecutor.java:54)
  at org.gradle.internal.buildtree.InitDeprecationLoggingActionExecutor.execute(InitDeprecationLoggingActionExecutor.java:62)
  at org.gradle.internal.buildtree.InitProblems.execute(InitProblems.java:36)
  at org.gradle.internal.buildtree.DefaultBuildTreeContext.execute(DefaultBuildTreeContext.java:40)
  at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.lambda$execute$0(BuildTreeLifecycleBuildActionExecutor.java:71)
  at org.gradle.internal.buildtree.BuildTreeState.run(BuildTreeState.java:60)
  at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.execute(BuildTreeLifecycleBuildActionExecutor.java:71)
  at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$3.call(RunAsBuildOperationBuildActionExecutor.java:61)
  at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$3.call(RunAsBuildOperationBuildActionExecutor.java:57)
  at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:209)
  at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
  at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
  at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
  at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:166)
  at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
  at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
  at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor.execute(RunAsBuildOperationBuildActionExecutor.java:57)
  at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.lambda$execute$0(RunAsWorkerThreadBuildActionExecutor.java:36)
  at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:263)
  at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:127)
  at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.execute(RunAsWorkerThreadBuildActionExecutor.java:36)
  at org.gradle.tooling.internal.provider.continuous.ContinuousBuildActionExecutor.execute(ContinuousBuildActionExecutor.java:110)
  at org.gradle.tooling.internal.provider.SubscribableBuildActionExecutor.execute(SubscribableBuildActionExecutor.java:64)
  at org.gradle.internal.session.DefaultBuildSessionContext.execute(DefaultBuildSessionContext.java:46)
  at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor$ActionImpl.apply(BuildSessionLifecycleBuildActionExecutor.java:92)
  at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor$ActionImpl.apply(BuildSessionLifecycleBuildActionExecutor.java:80)
  at org.gradle.internal.session.BuildSessionState.run(BuildSessionState.java:71)
  at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor.execute(BuildSessionLifecycleBuildActionExecutor.java:62)
  at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor.execute(BuildSessionLifecycleBuildActionExecutor.java:41)
  at org.gradle.internal.buildprocess.execution.StartParamsValidatingActionExecutor.execute(StartParamsValidatingActionExecutor.java:64)
  at org.gradle.internal.buildprocess.execution.StartParamsValidatingActionExecutor.execute(StartParamsValidatingActionExecutor.java:32)
  at org.gradle.internal.buildprocess.execution.SessionFailureReportingActionExecutor.execute(SessionFailureReportingActionExecutor.java:51)
  at org.gradle.internal.buildprocess.execution.SessionFailureReportingActionExecutor.execute(SessionFailureReportingActionExecutor.java:39)
  at org.gradle.internal.buildprocess.execution.SetupLoggingActionExecutor.execute(SetupLoggingActionExecutor.java:47)
  at org.gradle.internal.buildprocess.execution.SetupLoggingActionExecutor.execute(SetupLoggingActionExecutor.java:31)
  at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:70)
  at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
  at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
  at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
  at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
  at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
  at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
  at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
  at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
  at org.gradle.launcher.daemon.server.exec.ForwardClientInput.lambda$execute$0(ForwardClientInput.java:40)
  at org.gradle.internal.daemon.clientinput.ClientInputForwarder.forwardInput(ClientInputForwarder.java:80)
  at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:37)
  at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
  at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:64)
  at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
  at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
  at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
  at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
  at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:84)
  at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
  at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
  at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
  at org.gradle.launcher.daemon.server.DaemonStateCoordinator.lambda$runCommand$0(DaemonStateCoordinator.java:321)
  at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
  at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:48)
BUILD FAILED in 1s
Configuration cache entry discarded due to serialization error.
```

### Expected Behavior

No serialization errors

### Context (optional)

_No response_

### Self-contained Reproducer Project

As the issue is random, I'm unable to reproduce it reliably. The following combination of flags seems to cause the issue:

```properties
org.gradle.configuration-cache=true
org.gradle.configuration-cache.parallel=true
org.gradle.configureondemand=true
org.gradle.unsafe.isolated-projects=true
```

### Gradle version

8.11

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/subprojects/build-events/build.gradle.kts
+++ b/subprojects/build-events/build.gradle.kts
@@ -17,6 +17,8 @@ dependencies {
     implementation(projects.modelCore)
 
     api(libs.jsr305)
+
+    implementation(libs.errorProneAnnotations)
     implementation(libs.guava)
 
     testImplementation(projects.internalTesting)

--- a/subprojects/build-events/src/main/java/org/gradle/internal/build/event/DefaultBuildEventsListenerRegistry.java
+++ b/subprojects/build-events/src/main/java/org/gradle/internal/build/event/DefaultBuildEventsListenerRegistry.java
@@ -158,17 +158,17 @@ public class DefaultBuildEventsListenerRegistry implements BuildEventsListenerRe
     }
 
     private void unsubscribeAll() {
+        Collection<AbstractListener<?>> subscribed;
+
         synchronized (subscriptions) {
-            try {
-                Collection<AbstractListener<?>> subscribed = subscriptions.values();
-                subscribed.stream()
-                    .flatMap(it -> it.getListeners().stream())
-                    .forEach(this::unsubscribe);
-                CompositeStoppable.stoppable(subscribed).stop();
-            } finally {
-                subscriptions.clear();
-            }
+            subscribed = ImmutableList.copyOf(subscriptions.values());
+            subscriptions.clear();
         }
+
+        subscribed.stream()
+            .flatMap(it -> it.getListeners().stream())
+            .forEach(this::unsubscribe);
+        CompositeStoppable.stoppable(subscribed).stop();
     }
 
     private void unsubscribe(Object listener) {

--- a/subprojects/build-events/src/main/java/org/gradle/internal/build/event/DefaultBuildEventsListenerRegistry.java
+++ b/subprojects/build-events/src/main/java/org/gradle/internal/build/event/DefaultBuildEventsListenerRegistry.java
@@ -17,6 +17,7 @@
 package org.gradle.internal.build.event;
 
 import com.google.common.collect.ImmutableList;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
 import org.gradle.BuildAdapter;
 import org.gradle.BuildResult;
 import org.gradle.api.internal.GradleInternal;
@@ -61,11 +62,13 @@ import java.util.concurrent.Future;
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
 
 public class DefaultBuildEventsListenerRegistry implements BuildEventsListenerRegistry, BuildEventListenerRegistryInternal {
     private final BuildEventListenerFactory factory;
     private final ListenerManager listenerManager;
     private final BuildOperationListenerManager buildOperationListenerManager;
+    @GuardedBy("subscriptions")
     private final Map<Provider<?>, AbstractListener<?>> subscriptions = new LinkedHashMap<>();
     private final ExecutorFactory executorFactory;
 
@@ -84,7 +87,9 @@ public class DefaultBuildEventsListenerRegistry implements BuildEventsListenerRe
 
     @Override
     public List<Provider<?>> getSubscriptions() {
-        return ImmutableList.copyOf(subscriptions.keySet());
+        synchronized (subscriptions) {
+            return ImmutableList.copyOf(subscriptions.keySet());
+        }
     }
 
     @Override
@@ -99,31 +104,37 @@ public class DefaultBuildEventsListenerRegistry implements BuildEventsListenerRe
 
     @Override
     public void onOperationCompletion(Provider<? extends BuildOperationListener> listenerProvider) {
-        if (subscriptions.containsKey(listenerProvider)) {
-            return;
-        }
+        addSubscriptionIfNotExists(listenerProvider, this::makeBuildOperationSubscription);
+    }
 
+    private ForwardingBuildOperationListener makeBuildOperationSubscription(Provider<? extends BuildOperationListener> listenerProvider) {
         ForwardingBuildOperationListener subscription = new ForwardingBuildOperationListener(listenerProvider, executorFactory);
         processIfBuildService(listenerProvider);
-        subscriptions.put(listenerProvider, subscription);
         buildOperationListenerManager.addListener(subscription);
+        return subscription;
     }
 
     @Override
     public void onTaskCompletion(Provider<? extends OperationCompletionListener> listenerProvider) {
-        if (subscriptions.containsKey(listenerProvider)) {
-            return;
-        }
+        addSubscriptionIfNotExists(listenerProvider, this::makeTaskCompletionSubscription);
+    }
 
+    private ForwardingBuildEventConsumer makeTaskCompletionSubscription(Provider<? extends OperationCompletionListener> listenerProvider) {
         ForwardingBuildEventConsumer subscription = new ForwardingBuildEventConsumer(listenerProvider, executorFactory);
         processIfBuildService(listenerProvider);
-        subscriptions.put(listenerProvider, subscription);
 
         for (Object listener : subscription.getListeners()) {
             listenerManager.addListener(listener);
             if (listener instanceof BuildOperationListener) {
                 buildOperationListenerManager.addListener((BuildOperationListener) listener);
             }
+        }
+        return subscription;
+    }
+
+    private <T> void addSubscriptionIfNotExists(Provider<T> listenerProvider, Function<Provider<T>, ? extends AbstractListener<?>> factoryFunction) {
+        synchronized (subscriptions) {
+            subscriptions.computeIfAbsent(listenerProvider, Cast.uncheckedNonnullCast(factoryFunction));
         }
     }
 
@@ -136,7 +147,10 @@ public class DefaultBuildEventsListenerRegistry implements BuildEventsListenerRe
     }
 
     private void unsubscribeProvider(Provider<?> listenerProvider) {
-        AbstractListener<?> subscription = subscriptions.remove(listenerProvider);
+        AbstractListener<?> subscription;
+        synchronized (subscriptions) {
+            subscription = subscriptions.remove(listenerProvider);
+        }
         if (subscription != null) {
             subscription.getListeners().forEach(this::unsubscribe);
             subscription.close();
@@ -144,14 +158,16 @@ public class DefaultBuildEventsListenerRegistry implements BuildEventsListenerRe
     }
 
     private void unsubscribeAll() {
-        try {
-            Collection<AbstractListener<?>> subscribed = subscriptions.values();
-            subscribed.stream()
-                .flatMap(it -> it.getListeners().stream())
-                .forEach(this::unsubscribe);
-            CompositeStoppable.stoppable(subscribed).stop();
-        } finally {
-            subscriptions.clear();
+        synchronized (subscriptions) {
+            try {
+                Collection<AbstractListener<?>> subscribed = subscriptions.values();
+                subscribed.stream()
+                    .flatMap(it -> it.getListeners().stream())
+                    .forEach(this::unsubscribe);
+                CompositeStoppable.stoppable(subscribed).stop();
+            } finally {
+                subscriptions.clear();
+            }
         }
     }
 

